doc/examples/sharedLibraryInterfaceWithModesExample.cpp

```diff
@@ -0,0 +1,80 @@
+/* *****************************************************************************
+ * A.L.E (Arcade Learning Environment)
+ * Copyright (c) 2009-2013 by Yavar Naddaf, Joel Veness, Marc G. Bellemare,
+ *  Matthew Hausknecht, Marlos C. Machado and the Reinforcement Learning
+ *  and Artificial Intelligence Laboratory
+ * Released under the GNU General Public License; see License.txt for details. 
+ *
+ * Based on: Stella  --  "An Atari 2600 VCS Emulator"
+ * Copyright (c) 1995-2007 by Bradford W. Mott and the Stella team
+ *
+ * *****************************************************************************
+ *  sharedLibraryInterfaceExampleWithModes.cpp 
+ *
+ *  Sample code for running an agent with the shared library interface if one
+ *   wants to change the game mode or game difficult (available from v.0.6.0)
+ **************************************************************************** */
+
+#include <iostream>
+#include <ale_interface.hpp>
+
+#ifdef __USE_SDL
+  #include <SDL.h>
+#endif
+
+using namespace std;
+
+int main(int argc, char** argv) {
+    if (argc < 2) {
+        std::cerr << "Usage: " << argv[0] << " rom_file" << std::endl;
+        return 1;
+    }
+
+    ALEInterface ale;
+
+    // Get & Set the desired settings
+    ale.setInt("random_seed", 123);
+    //The default is already 0.25, this is just an example
+    ale.setFloat("repeat_action_probability", 0.25);
+
+#ifdef __USE_SDL
+    ale.setBool("display_screen", true);
+    ale.setBool("sound", true);
+#endif
+
+    // Load the ROM file. (Also resets the system for new settings to
+    // take effect.)
+    ale.loadROM(argv[1]);
+
+    // Get the vectors of available modes and difficulties
+    ModeVect modes = ale.getAvailableModes();
+    DifficultyVect difficulties = ale.getAvailableDifficulties();
+
+    cout << "Number of available modes: " << modes.size() << endl;
+    cout << "Number of available difficulties: " << difficulties.size() << endl;
+
+    // Get the vector of legal actions
+    ActionVect legal_actions = ale.getLegalActionSet();
+
+    // Play 10 episodes
+    for (int i = 0; i < modes.size(); i++){
+        for (int j = 0; j < difficulties.size(); j++){
+            
+            ale.setMode(modes[i]);
+            ale.setDifficulty(difficulties[j]);
```

Comment:
I'll look into the issues of resetting.
For the question of storage of difficutly and mode params, the reason was that the difficulty is something console-related (the physical position of the switches) while the mode is game-specific (which means custom change code).
Because of that, if we want to have both parameters stored in the same class, IMO it would make more sense in ROMSettings. But in any case, it would be weird somehow to have the parameter stored at one place, while the code that actually performs the modification is somewhere else.

About the serialization : do we want anyway to serialize these parameters ? I'm not sure to see the purpose.